docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+# Protocol and Chain Versions
+
+The Software Version is covered by SemVer and described elsewhere.
+It is not relevant to the protocol description, suffice to say that if any protocol version
+changes, the software version changes. Software version is included in NodeInfo for convenience/diagnostics.
+
+Here we focus on protocol and chain versions.
+
+## Requirements
+
+We need to version blockchains across protocols, networks, forks, etc.
+We need to do it in a way that is scalable and maintainable - we can't just litter
+the code with conditionals.
+We need chain identifiers and descriptions so we can talk about a multitude of chains in a meaningful way.
+
+Let's start with versions.
+
+### Protocol Version
+
+The complete version of the protocol contains the following sub-versions:
+BlockVersion, P2PVersion, AppVersion. These versions reflect the major sub-components
+of the software that are likely to evolve together, at different rates, and in different ways,
+as described below.
+
+#### BlockVersion
+
+- All tendermint hashed data-structures (headers, votes, txs, responses, etc.).
+	- Note the semantic meaning of a transaction may change according to the AppVersion,
+		but the way txs are merklized into the header is part of the BlockVersion
+- It should be the least frequent/likely to change.
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
+- To effectively change it, dump state and start a new blockchain with the new BlockVersion.
+- It's easy to determine the version of a block from its serialized form
+
+#### P2PVersion
+
+- All peer messaging (messages, detectable behaviour)
+- Will change gradually as reactors evolve to improve performance and support new features
+	- eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
+- It's easy to determine the version of a peer from its first serialized message/s
+- New versions must be compatible with at least one old version to allow gradual upgrades
+
+#### AppVersion
+
+- The ABCI state machine (txs, begin/endblock behaviour, commit hashing)
+- Behaviour and message types will change abruptly in the course of the life of a chain
+- Need to minimize complexity of the code for supporting different AppVersions at different heights
+- Ideally, each version of the software supports only a *single* AppVersion at one time
+    - this means we checkout different versions of the software at different heights instead of littering the code
```

Comment:
It would be *possible* to integrate Git and the Golang compiler into the SDK for actually "automatic" upgrades in this vein.

Not sure if it's worth the complexity (and not requiring manual upgrades introduces some risk).

Reply:
That would lock us into golang being the canonical language. I don't think we want that. (We should want other language implementations)

I also think having every update to golang or git on chain is a bad idea.

Reply:
I don't think that's true - it would make it *harder* to write implementations in other languages, but one could write the same kind of automation with another compiler (perhaps including in the upgrade two Git hashes for the different languages).

You would need a *lot* of protocol upgrades for the code size to be significant - at the moment, the SDK has ~8 MB of Go source (and it would be compressed in Git).

I agree that this is complex though.

Reply:
The source code can be compressed, which will cut that by a factor of 2-3 perhaps. 

Generally, though, I agree - this is definitely a post-launch consideration, and even then may not be worth the complexity.